File: LeetCode/00-legacy-warmups/find-the-index-of-the-first-occurrence-in-a-string/solutions/main_kmp.go
package main

func buildLPS(pattern string) []int {
	lps := make([]int, len(pattern))

	var length int = 0
	var i int = 1

	for i < len(pattern) {
		if pattern[i] == pattern[length] {
			length++
			lps[i] = length
			i++
		} else {
			if length > 0 {
				length = lps[length-1]
			} else {
				lps[i] = 0
				i++
			}
		}
	}

	return lps
}

func strStrKMP(heystask string, needle string) int {
	if len(needle) == 0 {
		return -1
	}

	// Build LPS (Longest Prefix Suffix) array

	var lps []int = buildLPS(needle)

	var i, j int = 0, 0

	for i < len(heystask) {
		if heystask[i] == needle[j] {
			i++
			j++
			if j == len(needle) {
				return i - j
			}
		} else {
			if j > 0 {
				j = lps[j-1]
			} else {
				i++
			}
		}
	}
	return -1
}

File: LeetCode/00-legacy-warmups/find-the-index-of-the-first-occurrence-in-a-string/solutions/main_kmp.js
function buildLPS(pattern) {
  const m = pattern.length;
  const lps = new Array(m).fill(0);
  let len = 0;
  let i = 1;

  while (i < m) {
    if (pattern[i] === pattern[len]) {
      len++;
      lps[i] = len;
      i++;
    } else {
      if (len !== 0) {
        len = lps[len - 1];
      } else {
        lps[i] = 0;
        i++;
      }
    }
  }
  return lps;
}

var strStr = function (haystack, needle) {
  const n = haystack.length;
  const m = needle.length;

  if (m === 0) return 0;
  if (n < m) return -1;

  const lps = buildLPS(needle);
  let i = 0;
  let j = 0;

  while (i < n) {
    if (haystack[i] === needle[j]) {
      i++;
      j++;

      if (j === m) {
        return i - j;
      }
    } else {
      if (j !== 0) {
        j = lps[j - 1];
      } else {
        i++;
      }
    }
  }
  return -1;
};
